### PR TITLE
Log when mismatched `referenceSequenceNumber` messages are detected by the batch manager - main

### PR DESCRIPTION
**Pull request by andre4i**
For compression to work as expected and maintain batch correctness, it is required that all messages within the same batch have the same reference sequence number.

The feature was partially ported from `next` (without the API on the runtime layer to block op creation from the event handler) and it is disabled by default (see https://github.com/microsoft/FluidFramework/pull/12326 and the description for details). **Compression would enable it.**

By default, the feature is disabled. However, we'd like to know if and when it happens, so this change will log at most 5 instances per session along with the stacktrace in order to help us better understand the customer scenarios and find a workaround.

Also fixing the issue with the AgentScheduler performing op reentry (also ported from `next`).

Regardless of the runtime options (including the compression options), when the feature is explicitly enabled, if the feature gate `Fluid.ContainerRuntime.DisableOpReentryCheck` is set to `true`, the feature will end up disabled.

### CODE REVIEW

**Comment by vladsud** [comment]
@andre4i - overall observation: We likely will need to enable all these features pretty soon in Bohemia code base. As such, validating that we do not have hidden work (product breaks because of these new limitations) will be extremely useful in short amount of time

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1038,6 +1049,13 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             this.validateSummaryHeuristicConfiguration(this.summaryConfiguration);
         }
 
+        this.enableOpReentryCheck = (runtimeOptions.enableOpReentryCheck === true
+            // If compression is enabled, we need to disallow op reentry as it is required that
+            // ops within the same batch have the same reference sequence number.
+            || runtimeOptions.compressionOptions.minimumBatchSizeInBytes !== Number.POSITIVE_INFINITY)
+            // Allow for a break-glass config to override the options
+            && this.mc.config.getBoolean("Fluid.ContainerRuntime.DisableOpReentryCheck") !== true;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1057]: Do we need runtimeOptions.enableOpReentryCheck? I.e. should we have fewer ways to enable same thing? Feature gate should be enough - it should be easy (if client wants so) to programmatically always set the settings (and overwrite control tower in code).

File: packages/runtime/container-runtime/src/opLifecycle/batchManager.ts [hunk 2]
```diff
@@ -87,6 +108,45 @@ export class BatchManager {
...
+            enableOpReentryCheck: this.options.enableOpReentryCheck === true,
+        };
+        const error = new GenericError(
+            "Submission of an out of order message",
+                /* error */ undefined,
+            telemetryProperties);
+
+        if (this.options.enableOpReentryCheck === true) {
+            this.logger.sendErrorEvent(
+                { eventName: "Submission of an out of order message", ...telemetryProperties },
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/batchManager.ts:132]: eventNames should not be English sentences.

File: packages/runtime/container-runtime/src/opLifecycle/batchManager.ts [hunk 3]
```diff
@@ -87,6 +108,45 @@ export class BatchManager {
...
+        }
+
+        const telemetryProperties = {
+            referenceSequenceNumber: this.pendingBatch[0].referenceSequenceNumber,
+            messageReferenceSequenceNumber: message.referenceSequenceNumber,
+            type: message.deserializedContent.type,
+            length: this.pendingBatch.length,
+            enableOpReentryCheck: this.options.enableOpReentryCheck === true,
+        };
+        const error = new GenericError(
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/batchManager.ts:125]: Should this be UsageError? Not sure, but it feels like it should be, as it's an error caused by user usage pattern.

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1057]: I think we do. There's one way to enable this, via runtime options. The feature gate is a break-glass way of disabling it via control tower.

File: packages/runtime/container-runtime/src/opLifecycle/batchManager.ts (hunk 3 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/batchManager.ts:125]: Good point

**Comment by andre4i** [comment]
> overall observation: We likely will need to enable all these features pretty soon in Bohemia code base. As such, validating that we do not have hidden work (product breaks because of these new limitations) will be extremely useful in short amount of time

@vladsud 

I have checked that the bohemia repository does not break any tests if we just enable this, so I am cautiously optimistic about it.

**Pull request merged**